### ISSUE TITLE
add support for Predictive Coding as alternative to Backpropagation

### ISSUE THREAD

**Issue by MovGP0** [pytorch/pytorch#177623]
### 🚀 The feature, motivation and pitch

There is some evidence that [Predictive Coding](https://en.wikipedia.org/wiki/Predictive_coding) (PC) can outperform gradient estimation over Backpropagation (BP) in some use cases:

- online learning
- training on small datasets
- continual learning
- possible energy efficiency gains through localized error estimation

Thus, we should consider to implement this algorithm into torchlib as an alternative to BP.

### Alternatives

**Backpropagation**

There has been more research done in BP, so PB currently outperforms PC on most tasks. 

(Alternatives to) algorithms like [Batch Normalization](https://arxiv.org/abs/1502.03167), Optimizers like [ADAM](https://arxiv.org/abs/1412.6980), and [Dropout](https://www.jmlr.org/papers/volume15/srivastava14a/srivastava14a.pdf) do not exist (yet) for PC.

> My argument is that we should consider implementing it anyway, such that those algorithms can be developed.

### Papers
- [Predictive Coding: Towards a Future of Deep Learning beyond Backpropagation](https://arxiv.org/abs/2202.09467)
- [The neural coding framework for learning generative models](https://www.nature.com/articles/s41467-022-29632-7)
- [Learning multiple layers of representation](https://www.cell.com/trends/cognitive-sciences/abstract/S1364-6613(07)00217-3)

### Talk
- [On the past, present, and future of predictive coding](https://www.youtube.com/watch?v=MKmN9ov1Wsc)

cc @ezyang @albanD @gqchen @nikitaved @soulitzer @Varal7 @bobrenjc93 @vincentqb @jbschlosser @janeyx99 @crcrpar
